(86) 'Back' links throughout the form

dgmstuart opened the pull request:
**Merge https://github.com/LBHackney-IT/maintain-my-home/pull/17 first**

The first step of this was to glue all the pages together so that we can now click through the form (nearly) end-to-end and back again.

There's a lot of duplication in the specs - it would be nice to get rid of that somehow, but although it's currently possible to jump to arbitrary points in the form (which would be one way to shortcut this), that won't be true in future (when we check what questions you've already answered), so I don't want to cause us future pain by creating specs which will break later.

This is currently hardcoded per controller. We'll need to do something special for branching and rejoining flows, but currently there aren't any through the app (e.g. there's no way to get to the unknown repair description page) so I propose we deal with that later and implement the pending specs which I've left at that point.

In any case - there's already a LOT here :scream:

dgmstuart committed:
(feature) Residents can navigate back in the form

Go through each page as simply as possible and then click on all the
back links all the way back to the start page

The pending specs are there to remind us to implement the back links on
the less obvious pages.

Diagnosis pages are not implemented yet, so skip that page for now.

In this initial implementation we hardcode the links into the page -
we'll make that more flexible in the next step.

dgmstuart committed:
(feature) Presenter class for 'back' links

These links are going to be on multiple pages so we need a consistent
way of generating them.

This approach assumes that we'll always be redirecting back to the index
action of some controller

The test is a bit fiddly since basically we're just composing bits from
I18n and the route helpers. We'd get that for free if we did this as a
helper, but then we'd have to pass some information to the view which
the view probably shouldn't know, so I prefer to have this in an object
which can be built in the controller.

dgmstuart committed:
(spec) go back when address search invalid

This currently passes with a hardcoded implementation, but will fail
when we start passing the information from the controller

dgmstuart committed:
(refactor) use the Back presenter in views

dgmstuart committed:
(feature) Emergency contact page has a back link

dgmstuart committed:
(feature) Address Not Here page has a back link

dgmstuart committed:
(feature) Back link on the address selection page

Note that both pages go back to problem description, as per the
prototype: the 'change' link takes you back to address search, so that's
already covered.

Note that previously we were generating Back only for the index page,
and now we're generating the same object for both branches

leeky commented:
Rebasing

leeky commented on app/presenters/back.rb:
```diff
@@ -0,0 +1,25 @@
+  private
+
+  def text
+    "#{prefix} #{page_description}"
```
You could also generate this text using I18n interpolation, e.g in `en.yml`:

`Back to %{page_description}`

Which is invaluable if we were translating to a language where "Back to" should be last in the sentence.

dgmstuart replied:
Good shout. I was wondering about that. I guess it makes no difference that the parameter passed to it will be another I18n string.

dgmstuart replied:
@leeky pushed a commit for this

dgmstuart committed:
(refactor) string interpolation in backlink prefix

This allows translations where 'back to' would be at the end of the
string.

The stubbing in the spec is now a little nonsensical, but I'd prefer to
have it this way than have the spec rely on en.yml having specific
strings in it.

leeky reviewed (approved):
LGTM!

@dgmstuart Could you rebase, and then I'm happy for this to be merged.

dgmstuart committed:
(feature) back link from invalid contact details

When the contact details form has a validation error we need to make
sure it still has a back link on it.

dgmstuart commented:
A sadly painful rebase. Ready for another look. One new commit (2c961cb) and amended 7cd9e8b to add in the `@back` object in the case where the AddressForm is invalid. 

On the plus side, the existing tests told us something was wrong here, which gives me even more confidence in our test suite.

Ready for another look
